### ISSUE TITLE
Search clarification

### ISSUE THREAD

**Issue by trondhindenes** [jaegertracing/jaeger#1568]
## Requirement - what kind of business use case are you trying to solve?
I'm not sure if this is a bug, but it would still be great to get some feedback on it. Trying to describe what we're seeing as best I can in the problem section below

## Problem - what in Jaeger blocks you from solving the requirement?
During load-testing of one of our apis I noticed the following:
The load test produced a lot of spans, let's say 500 per second. A high percentage of these failed, and had `error=true` tagged on one or multiple traces in the span.
We're using Elasticsearch 6.x as our storage backend.

Here's the strange part: if I search for tag `error=true` with `limit results=1000` I would only get a hit if any of the last 1000 traces had error in it. So it seems to me that Jaeger first limits the "searchable blocks" and then applies tag filters etc. 

I would argue that the search should work by _first_ applying any custom tag filters etc, and then if the resulting dataset is bigger than `limit results`, limit the returned data. 

Since Jaeger has an upper limit of 1500 records, we were simply unable to find the spans with errors in them produced by our load testing, despite knowing that they were there.

## Proposal - what do you suggest to solve the problem or improve the existing situation?
I guess for one it would be good to get some confirmation on how jaeger actually performs searches against Elasticsearch, and whether or not our experiences was expected or not (e.g. whether or not the `limit results` value is applied before or after filtering.

**Comment by yurishkuro** [comment]
to my knowledge, there is no sequential logic in the search, Jaeger sends a single query that specifies the time range, tag expressions, and the limit.

**Comment by pavolloffay** [comment]
@trondhindenes what version of Jaeger are you using?

**Comment by trondhindenes** [comment]
we're on 1.12 now, but I noticed this before the upgrade, so 1.11.

**Comment by trondhindenes** [comment]
thanks @yurishkuro I guess I can poke around in how Jaeger performes queries against ES to get to the bottom of this.

**Comment by trondhindenes** [comment]
So I poked a bit more by placing a Charles proxy between Jaeger-Query and Elasticsearch. This is a screenshot of Jaeger's queries against the  `msearch` api:
![Screenshot from 2019-07-01 09-34-24](https://user-images.githubusercontent.com/1747120/60418588-961ee780-9be3-11e9-94ba-2b33e668f828.png)

I think the "terminate after" is what causes the weird behavior we're seeing. As documented here, https://www.elastic.co/guide/en/elasticsearch/reference/7.1/search-request-body.html, Elasticsearch will simply "hard drop" the query when it reaches the `terminate after` document count. This means that if I have 1 trace with `error=true` followed by 10000 traces where `error=false`, I'll never be able to find the "error trace" in the search.

**Comment by trondhindenes** [comment]
Possibly a separate issue, but as far as I can see from this, it looks to me as queries such as `error=true` are not sent to Elasticsearch - they seem to be applied "client-side" after Jaeger Query has fetched all documents. This strikes me as an inefficient design (there might be reasons for it that I'm not aware of tho)

**Comment by yurishkuro** [comment]
the documentation for "terminate after" says "`terminate_after` is always applied after the `post_filter`". A further example says "Also we can set terminate_after to 1 to indicate that the query execution can be terminated whenever the first matching document was found (per shard)." That seems to indicate that "terminate after" applies to _matching results_, not to the raw number of documents scanned.

**Comment by yurishkuro** [comment]
> it looks to me as queries such as error=true are not sent to Elasticsearch

yeah, that would be strange (a bug)

**Comment by trondhindenes** [comment]
> that seems to indicate that "terminate after" applies to matching results
you're right. Still, the fact that `error=true` doesn't seem to be sent to elasticsearch means that we'll likely just get the `n` newest results back. There's a `--es.tags-as-fields.config-file string` parameter that might be related, but I struggle to find any documentation about how to use it or what it's for.

**Comment by trondhindenes** [comment]
Adding some more to this as I go:
Jaeger's doc format in ES might look something like this:

```json
{
  "_index": "jaeger-span-2019-07-03",
  "_type": "span",
  "_id": "8osIuWsBiOrdj4b7p5an",
  "_score": 1,
  "_source": {
    "traceID": "59e3f22acedccd25",
    "spanID": "127bcf03bdc6d8c2",
    "flags": 1,
    "operationName": "findTraceIDs",
    "references": [
      {
        "refType": "CHILD_OF",
        "traceID": "59e3f22acedccd25",
        "spanID": "5c49cf5d05ac8e21"
      }
    ],
    "startTime": 1562177479574203,
    "startTimeMillis": 1562177479574,
    "duration": 26198,
    "tags": [
      {
        "key": "internal.span.format",
        "type": "string",
        "value": "proto"
      }
    ],
    "logs": [],
    "process": {
      "serviceName": "jaeger-query",
      "tags": [
        {
          "key": "jaeger.version",
          "type": "string",
          "value": "Go-2.15.1dev"
        },
        {
          "key": "hostname",
          "type": "string",
          "value": "trond-x1"
        },
        {
          "key": "ip",
          "type": "string",
          "value": "10.90.2.52"
        },
        {
          "key": "client-uuid",
          "type": "string",
          "value": "581eebffa0935ece"
        }
      ]
    }
  },
  "fields": {
    "startTimeMillis": [
      "2019-07-03T18:11:19.574Z"
    ]
  }
}
```

There's multiple attributes containing "tags", and jaeger-query seems to be able to search all of them (I can search for tags in both `process.tags` and `tags`)

If I wanted to use "server-side" searching, I'd have to issue a `nested` type of search because of the way the documents are organized, something like
```json
{
  "query": {
    "nested": {
      "path": "tags",
      "query": {
        "bool": {
          "must": [
            { "match": { "tags.key": "error" }},
            { "match": { "tags.value": "true" }}
          ]
        }
      }
    }
  }
}
```
all "paths" containing tags would have to be searched in order to provide a complete result (for example, in order to search for the "hostname" field which is stored in `process.tags`, I'd have to adjust the `path` of the query to `process.tags`. 

All in all, its not trivial but I'm increasingly getting the feeling that there's tons of optimizations that could be done around how Jaeger stores data in Elasticsearch.

**Comment by trondhindenes** [comment]
@yurishkuro sorry to ping. I just looked up on this. It would be good to get it confirmed that I understand the behavior correctly:

> queries such as error=true are not sent to Elasticsearch - they seem to be applied "client-side" after Jaeger Query has fetched all documents

This poses a problem for us, for the reasons I've described. I totally understand that changing the way Jaeger queries data in ES is probably a complex endeavour. We'd probably evaluate Cassandra instead to see if that works better for us. In any case, it would be good to get the statement confirmed.

**Comment by yurishkuro** [comment]
I am not aware of any client-side filtering, all search params should be sent to ES. What makes you think otherwise?

cc @pavolloffay

**Comment by trondhindenes** [comment]
as I write further up, I traced the requests sent from Jaeger to ES. It would explain why we didn't get any hits in our `error=true` query which lead me to open this issue in the first place.

**Comment by pavolloffay** [comment]
I am not aware of any client-side filtering the match for tag key and values is done in the Elasticsearch.

**Comment by trondhindenes** [comment]
thanks. I'll re-test my assumptions.
